File: runtime/Java/src/org/antlr/v4/runtime/misc/MultiMap.java
/*
 * Copyright (c) 2012-2017 The ANTLR Project. All rights reserved.
 * Use of this file is governed by the BSD 3-clause license that
 * can be found in the LICENSE.txt file in the project root.
 */

package org.antlr.v4.runtime.misc;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;

public class MultiMap<K, V> extends LinkedHashMap<K, List<V>> {
	public void map(K key, V value) {
		List<V> elementsForKey = get(key);
		if ( elementsForKey==null ) {
			elementsForKey = new ArrayList<V>();
			super.put(key, elementsForKey);
		}
		elementsForKey.add(value);
	}

	public List<Pair<K,V>> getPairs() {
		List<Pair<K,V>> pairs = new ArrayList<Pair<K,V>>();
		for (K key : keySet()) {
			for (V value : get(key)) {
				pairs.add(new Pair<K,V>(key, value));
			}
		}
		return pairs;
	}
}

File: runtime/Swift/Sources/Antlr4/misc/MultiMap.swift
/// Copyright (c) 2012-2016 The ANTLR Project. All rights reserved.
/// Use of this file is governed by the BSD 3-clause license that
/// can be found in the LICENSE.txt file in the project root.


public class MultiMap<K:Hashable, V> {
    private var mapping = [K: Array < V>]()
    public func map(_ key: K, _ value: V) {
        var elementsForKey: Array<V>? = mapping[key]
        if elementsForKey == nil {
            elementsForKey = Array<V>()
            mapping[key] = elementsForKey
        }
        elementsForKey?.append(value)
    }

    public func getPairs() -> Array<(K, V)> {
        var pairs: Array<(K, V)> = Array<(K, V)>()
        for key: K in mapping.keys {
            for value: V in mapping[key]! {
                pairs.append((key, value))
            }
        }
        return pairs
    }

    public func get(_ key: K) -> Array<(V)>? {
        return mapping[key]
    }

    public func size() -> Int {
        return mapping.count
    }

}
